crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dt", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dat", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/ds", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/e", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("s/dc/#", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+        cloud_client
+            .subscribe("error", QoS::AtLeastOnce)
+            .await
+            .unwrap();
+
+        let local_to_cloud = tokio::spawn(async move {
+            loop {
+                let notification = match cloud_event_loop.poll().await {
+                    Ok(notification) => notification,
+                    Err(err) => {
+                        error!("MQTT bridge connection error: {err}");
+                        continue;
+                    }
+                };
+                if let Event::Incoming(Incoming::Publish(publish)) = &notification {
+                    local_client
+                        .publish(
+                            format!("{prefix}/{}", publish.topic),
+                            publish.qos,
+                            publish.retain,
+                            publish.payload.clone(),
+                        )
+                        .await
+                        .unwrap();
+                    local_client.ack(publish).await.unwrap();
+                }
+            }
+        });
+        let cloud_to_local = tokio::spawn(async move {
+            let topic_prefix = format!("{prefix}/");
+            let mut last_error = None;
+            loop {
+                let notification = match local_event_loop.poll().await {
+                    Ok(notification) => {
+                        if let Event::Incoming(Incoming::ConnAck(_)) = &notification {
+                            info!("MQTT bridge connection re-established");
+                        }
+                        notification
+                    }
+                    Err(err) => {
+                        let current_error = err.to_string();
+                        // TODO repeat log after a timeout
+                        if last_error.as_ref() != Some(&current_error) {
+                            error!("MQTT bridge connection error: {err}");
+                        }
+                        last_error = Some(current_error);
+                        continue;
+                    }
+                };
+                if let Event::Incoming(Incoming::Publish(publish)) = &notification {
+                    cloud_client
+                        .publish(
+                            publish.topic.strip_prefix(&topic_prefix).unwrap(),
```

Comment:
Rather than some mapping hard-coded inside the bridge logic, I would consider to pass a function that convert a source topic to a target topic.

Reply:
Shouldn't we consider externalising the mapping rules, where the mapping rules can be defined in an external config file(similar to what Mosquitto has done) and use that for the mapping? That way, we can reuse the same bridge actor with all the cloud mappers using diff config files. It also enables the user to extend the mapper with additional mapping rules for things like custom templates.

Reply:
We need to think carefully on this. We don't need to make the mapping too open 

1. Mappings which are critical to the function of the agent (e.g. `c8y/s/ds` and `c8y/s/us` etc.)
2. Optional user specified mapping (e.g. `c8y/dc/<TemplateCollectionID>` etc.)

The first time of mappings should not be influenced by the user as this would lead to a non-functional agent

The second point is already configurable via the tedge.toml file, so we don't need a second level of configuration which converts the tedge.toml to the a mapping file rule. We could still open up to allow users to subscribe/publish additional topics (this will be important for the Cumulocity IoT MQTT Connect feature)...though we can still review the configuration mechanisms for the custom SmartREST templates etc.

Reply:
> The first time of mappings should not be influenced by the user as this would lead to a non-functional agent

Yeah definitely, the core and extended mapping rules must be separate to avoid the user messing up the core rules. How about keeping the core mapping rules in a separate file, which should not be directly manipulated by the user for any extensions?

Keeping the mapping rules outside the rust code makes the development of a new mapper easier (without even needing to know Rust), right? The goal was to achieve both "mqtt broker independence" and "generic mapping" with a single design.

Reply:
In my mind, the primary reason for the existence of the mapping rules we have in any case is so tedge-mapper can function correctly. It therefore seems a bit daft to me to store these outside of the mapper code (though they will need relocating to cope with non-Cumulocity clouds).

Additionally, I think the main hinderance to "generic mapping" at the moment is merely the fact it requires additional effort at the moment for someone to update mosquitto configurations, and so long as we don't need to do that, I don't think we're likely to introduce significant barriers to that being a possibility in the future. In general, it seems like a bad idea to try and attack both replacing the existing bridge configuration with something equivalent in the mapper and adding new features to the bridge (beyond trivial matters like automatically subscribing the bridge to new custom smartrest topics).

Reply:
> In my mind, the primary reason for the existence of the mapping rules we have in any case is so tedge-mapper can function correctly. It therefore seems a bit daft to me to store these outside of the mapper code (though they will need relocating to cope with non-Cumulocity clouds).

Externalizing the bridge rules is not needed if we want to keep things in the current form, with dedicated mapper processes for each cloud. But, with the desire to have a "generic mapper" eventually, which I'd assume will cater to all the clouds, an externalized config file seemed like the easiest way, especially to enable no-code development of mappers.

> But we don't need an intermediate file format...we only used that in the past as it was required by mosquitto, but now we are in control if it, then we don't need it.

Good point. This makes me wonder if we should reconsider the current mapping/bridging approach as well. So far, we needed the `c8y/#` and `az/#` like bridge topics, just because the bridge was external. So, the mapper was forced to send the mapped messages to the broker so that the external bridge client can pick them up. But now that we have the converter and the bridge in the same process, the messages between the converter actor and the bridge actor can be exchanged via in-memory channels (message boxes) rather than the broker, eliminating all the serializing/deserializing cost + network hops via the broker.

Obviously, this would be a much bigger change and the approach taken in this PR can be considered as the first step towards that goal.

> We still need to decide what the generic mapping looks like, as I think it is too early to tell if this is configuration, templates or code modules.

Fair point. I was just trying to see if there's a common ground here.

In my head, the generic mapper is something that behaves like(**not directly using**) the mosquitto bridge, but with a more sophisticated mapping rule-set for topic/payload manipulations as well beyond the simple topic-to-topic bridging offered by mosquitto. So, an external config file with just the bridge config aspects felt like a good first step, which can later be enhanced with the relevant mapping rules as well.

But, having this conversation, I get a feeling that we all have different views on how that generic mapper would work, and hence may be it's too early to discuss those aspects under the scope of this ticket.